cpp/src/arrow/array/builder_union.cc

```diff
@@ -19,42 +19,129 @@
 
 #include <utility>
 
+#include "arrow/util/checked_cast.h"
 #include "arrow/util/logging.h"
 
 namespace arrow {
 
+using internal::checked_cast;
+
+DenseUnionBuilder::DenseUnionBuilder(MemoryPool* pool)
+    : ArrayBuilder(nullptr, pool), types_builder_(pool), offsets_builder_(pool) {}
+
 DenseUnionBuilder::DenseUnionBuilder(MemoryPool* pool,
+                                     std::vector<std::shared_ptr<ArrayBuilder>> children,
                                      const std::shared_ptr<DataType>& type)
-    : ArrayBuilder(type, pool), types_builder_(pool), offsets_builder_(pool) {}
+    : ArrayBuilder(type, pool), types_builder_(pool), offsets_builder_(pool) {
+  auto union_type = checked_cast<const UnionType*>(type.get());
+  DCHECK_NE(union_type, nullptr);
+  type_id_to_child_num_.resize(union_type->max_type_code() + 1, -1);
+  DCHECK_EQ(union_type->mode(), UnionMode::DENSE);
+  int child_i = 0;
+  for (auto type_id : union_type->type_codes()) {
```

Comment:
I can let `children_[i]` be the builder for `type_id == i` instead of the builder which will finish into `child_data[i]`. The latter seems like it might be an implicit contract though; I'll investigate